File: Project/Simulator/Organisms/Animals/Human.java
package Simulator.Organisms.Animals;

import Simulator.Organisms.Organism;
import Simulator.GlobalSettings;
import Utils.DynamicDirections;
import Simulator.Abilities.Immortality;
import Simulator.World;

import org.json.JSONObject;

public class Human extends Animal {
    private DynamicDirections nextMove = DynamicDirections.get("SELF");
    private Immortality immortality;

    public Human(World world) {
        super(5, 4, world, Type.HUMAN);
        this.immortality = new Immortality(5, 5);
    }

    public Human(JSONObject json, World world) {
        super(json, world);
        this.immortality = new Immortality(json.getInt("ability_cooldown"), json.getInt("immortality_left"), 5, 5);
        if (json.getInt("immortality_left") > 0 || json.getInt("ability_cooldown") == 0) {
            immortality.use();
        }
    }

    @Override
    public Animal construct() {
        return new Human(world);
    }

    @Override
    public void action() {
        if (GlobalSettings.HUMAN_AI) {
            immortality.use();
            immortality.update();
            super.action();
            return;
        }
        if (nextMove == DynamicDirections.get("SELF"))
            return;
        if (immortality.checkToggle()) {
            immortality.use();
            immortality.flipToggle();
        }
        immortality.update();
        move(nextMove);
        nextMove = DynamicDirections.get("SELF");
    }

    @Override
    public void collision(Organism other) {
        if (immortality.isActive()) {
            immortality.effect(this, other);
            return;
        } else
            super.collision(other);
    }

    @Override
    public boolean collisionReaction(Organism other) {
        if (immortality.isActive()) {
            immortality.effect(this, other);
            return true;
        } else
            return false;
    }

    public void setNextMove(DynamicDirections direction) {
        this.nextMove = direction;
    }

    public void toggleImmortality() {
        if (immortality.isReady() && !immortality.isActive())
            immortality.flipToggle();
    }

    public DynamicDirections getNextMove() {
        return nextMove;
    }

    public String getAbilityInfo() {
        if (immortality.isActive())
            return immortality.getDuration() + " turns left";
        if (!immortality.isReady())
            return immortality.getCooldown() + " turns of cooldown left";
        if (immortality.checkToggle())
            return "Using next turn";
        return "Ready to use";
    }

    @Override
    public void die() {
        if (immortality.isActive()) {
            immortality.effect(this);
        } else {
            super.die();
        }
    }

    public JSONObject toJson() {
        JSONObject json = super.toJson();
        json.put("ability_cooldown", immortality.getCooldown());
        json.put("immortality_left", immortality.getDuration());
        return json;
    }
}

File: Project/Simulator/Organisms/Animals/Human.py
from .Animal import Animal
from Simulator.Organisms.Organism import Type
from GlobalSettings import GlobalSettings
from Utils.DynamicDirections import DynamicDirections


class Human(Animal):
    def __init__(self, world, json=None):
        from Simulator.Abilities.Immortality import Immortality
        super().__init__(5, 4, world, Type.HUMAN, json)
        if (json is None):
            self.__immortality = Immortality(5, 5)
        else:
            self.__immortality = Immortality(
                json["ability_cooldown"], json["immortality_left"], 5, 5)
            if (json["immortality_left"] > 0 or json["ability_cooldown"] == 0):
                self.__immortality.use()
        self.__nextMove = DynamicDirections.get("SELF")

    def construct(self):
        return Human(self._world)

    def action(self):
        if (GlobalSettings.HUMAN_AI):
            self.__immortality.use()
            self.__immortality.update()
            super().action()
            return
        if (self.__nextMove == DynamicDirections.get("SELF")):
            return
        if (self.__immortality.checkToggle()):
            self.__immortality.use()
            self.__immortality.flipToggle()
        self.__immortality.update()
        self.move(self.__nextMove)
        self.__nextMove = DynamicDirections.get("SELF")

    def collision(self, other):
        if (self.__immortality.isActive()):
            self.__immortality.effect(self, other)
            return
        else:
            super().collision(other)

    def collisionReaction(self, other):
        if (self.__immortality.isActive()):
            self.__immortality.effect(self, other)
            return True
        else:
            return False

    def setNextMove(self, direction):
        self.__nextMove = direction

    def toggleImmortality(self):
        if (self.__immortality.isReady() and not self.__immortality.isActive()):
            self.__immortality.flipToggle()

    def getNextMove(self):
        if (self.__nextMove == DynamicDirections.get("SELF")):
            return "Give direction"
        return self.__nextMove

    def getAbilityInfo(self):
        if (self.__immortality.isActive()):
            return f"{self.__immortality.getDuration()} turns left"
        if (not self.__immortality.isReady()):
            return f"{self.__immortality.getCooldown()} turns of cooldown left"
        if (self.__immortality.checkToggle()):
            return "Using next turn"
        return "Ready to use"

    def die(self):
        if (self.__immortality.isActive()):
            self.__immortality.effect(self)
        else:
            super().die()

    def toJson(self):
        json = super().toJson()
        json["ability_cooldown"] = self.__immortality.getCooldown()
        json["immortality_left"] = self.__immortality.getDuration()
        return json
